{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "WZIOe2CLHBwo"
      },
      "outputs": [],
      "source": [
        "\n",
        "Data Model - PNT2022TMID40157\n",
        "Team leader - John Anthony V Team Member 1 - Ashok G Team Member 2 - Sriram E Team Member 3 - Kalaivani S\n",
        "\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "import warnings\n",
        "from sklearn.model_selection import KFold, StratifiedKFold, cross_val_score,train_test_split\n",
        "from sklearn import linear_model, tree, ensemble\n",
        "from sklearn.metrics import confusion_matrix,plot_confusion_matrix,classification_report\n",
        "from sklearn.model_selection import GridSearchCV\n",
        "dataframe=pd.read_csv(\"E:/7th sem/ibm/Heart_Disease_Prediction.csv\")\n",
        "dataframe.head(10)\n",
        "Age\tSex\tChest pain type\tBP\tCholesterol\tFBS over 120\tEKG results\tMax HR\tExercise angina\tST depression\tSlope of ST\tNumber of vessels fluro\tThallium\tHeart Disease\n",
        "0\t70\t1\t4\t130\t322\t0\t2\t109\t0\t2.4\t2\t3\t3\tPresence\n",
        "1\t67\t0\t3\t115\t564\t0\t2\t160\t0\t1.6\t2\t0\t7\tAbsence\n",
        "2\t57\t1\t2\t124\t261\t0\t0\t141\t0\t0.3\t1\t0\t7\tPresence\n",
        "3\t64\t1\t4\t128\t263\t0\t0\t105\t1\t0.2\t2\t1\t7\tAbsence\n",
        "4\t74\t0\t2\t120\t269\t0\t2\t121\t1\t0.2\t1\t1\t3\tAbsence\n",
        "5\t65\t1\t4\t120\t177\t0\t0\t140\t0\t0.4\t1\t0\t7\tAbsence\n",
        "6\t56\t1\t3\t130\t256\t1\t2\t142\t1\t0.6\t2\t1\t6\tPresence\n",
        "7\t59\t1\t4\t110\t239\t0\t2\t142\t1\t1.2\t2\t1\t7\tPresence\n",
        "8\t60\t1\t4\t140\t293\t0\t2\t170\t0\t1.2\t2\t2\t7\tPresence\n",
        "9\t63\t0\t4\t150\t407\t0\t2\t154\t0\t4.0\t2\t3\t7\tPresence\n",
        "x = dataframe.iloc[ : , :-1].values\n",
        "y = dataframe.iloc[ : , -1].values\n",
        "X_train, X_test,y_train, y_test=train_test_split(x,y,test_size=0.25,random_state=40)\n",
        "from sklearn.model_selection import RandomizedSearchCV\n",
        "from sklearn.tree import DecisionTreeClassifier\n",
        "\n",
        "tree_model = DecisionTreeClassifier(max_depth=5,criterion='entropy')\n",
        "cv_scores = cross_val_score(tree_model, x, y, cv=10, scoring='accuracy')\n",
        "m=tree_model.fit(x, y)\n",
        "prediction=m.predict(X_test)\n",
        "cm= confusion_matrix(y_test,prediction)\n",
        "sns.heatmap(cm, annot=True,cmap='winter',linewidths=0.3, linecolor='black',annot_kws={\"size\": 20})\n",
        "print(classification_report(y_test, prediction))\n",
        "\n",
        "TP=cm[0][0]\n",
        "TN=cm[1][1]\n",
        "FN=cm[1][0]\n",
        "FP=cm[0][1]\n",
        "print('Testing Accuracy for Decision Tree:',(TP+TN)/(TP+TN+FN+FP))\n",
        "print('Testing Sensitivity for Decision Tree:',(TP/(TP+FN)))\n",
        "print('Testing Specificity for Decision Tree:',(TN/(TN+FP)))\n",
        "print('Testing Precision for Decision Tree:',(TP/(TP+FP)))\n",
        "              precision    recall  f1-score   support\n",
        "\n",
        "     Absence       0.87      1.00      0.93        40\n",
        "    Presence       1.00      0.79      0.88        28\n",
        "\n",
        "    accuracy                           0.91        68\n",
        "   macro avg       0.93      0.89      0.91        68\n",
        "weighted avg       0.92      0.91      0.91        68\n",
        "\n",
        "Testing Accuracy for Decision Tree: 0.9117647058823529\n",
        "Testing Sensitivity for Decision Tree: 0.8695652173913043\n",
        "Testing Specificity for Decision Tree: 1.0\n",
        "Testing Precision for Decision Tree: 1.0\n",
        "\n",
        "from sklearn.model_selection import RandomizedSearchCV\n",
        "from sklearn.tree import DecisionTreeClassifier\n",
        "parameters={\n",
        "    'criterion':('gini','entropy'),\n",
        "    'max_depth':(3,5,7,9,10),\n",
        "    'max_features':('auto','sqrt','log2'),\n",
        "    'min_samples_split': (2,4,6)\n",
        "}\n",
        "dt_grid = RandomizedSearchCV(DecisionTreeClassifier(), param_distributions = parameters, cv=10, verbose=True)\n",
        "dt_grid.fit(X_train,y_train)\n",
        "Fitting 10 folds for each of 10 candidates, totalling 100 fits\n",
        "RandomizedSearchCV(cv=10, estimator=DecisionTreeClassifier(),\n",
        "                   param_distributions={'criterion': ('gini', 'entropy'),\n",
        "                                        'max_depth': (3, 5, 7, 9, 10),\n",
        "                                        'max_features': ('auto', 'sqrt',\n",
        "                                                         'log2'),\n",
        "                                        'min_samples_split': (2, 4, 6)},\n",
        "                   verbose=True)\n",
        "dt_grid.best_estimator_\n",
        "DecisionTreeClassifier(max_depth=7, max_features='sqrt', min_samples_split=6)\n",
        "kf =KFold(n_splits=10, shuffle=True, random_state=42)\n",
        "\n",
        "cnt = 1\n",
        "for train_index, test_index in kf.split(x, y):\n",
        "    print(f'Fold:{cnt}, Train set: {len(train_index)}, Test set:{len(test_index)}')\n",
        "    cnt += 1\n",
        "Fold:1, Train set: 243, Test set:27\n",
        "Fold:2, Train set: 243, Test set:27\n",
        "Fold:3, Train set: 243, Test set:27\n",
        "Fold:4, Train set: 243, Test set:27\n",
        "Fold:5, Train set: 243, Test set:27\n",
        "Fold:6, Train set: 243, Test set:27\n",
        "Fold:7, Train set: 243, Test set:27\n",
        "Fold:8, Train set: 243, Test set:27\n",
        "Fold:9, Train set: 243, Test set:27\n",
        "Fold:10, Train set: 243, Test set:27\n",
        "def rmse(score):\n",
        "    rmse = np.sqrt(-score)\n",
        "    print(f'rmse= {\"{:.2f}\".format(rmse)}')\n",
        "score = cross_val_score(tree.DecisionTreeRegressor(random_state= 42), x, y, cv=kf, scoring=\"neg_mean_squared_error\")\n",
        "print(f'Scores for each fold: {score}')\n",
        "rmse(score.mean())\n",
        "Scores for each fold: [nan nan nan nan nan nan nan nan nan nan]\n",
        "rmse= nan\n",
        "C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:372: FitFailedWarning: \n",
        "10 fits failed out of a total of 10.\n",
        "The score on these train-test partitions for these parameters will be set to nan.\n",
        "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
        "\n",
        "Below are more details about the failures:\n",
        "--------------------------------------------------------------------------------\n",
        "9 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Presence'\n",
        "\n",
        "--------------------------------------------------------------------------------\n",
        "1 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Absence'\n",
        "\n",
        "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
        "max_depth = [1,2,3,4,5,6,7,8,9,10]\n",
        "\n",
        "for val in max_depth:\n",
        "    score = cross_val_score(tree.DecisionTreeRegressor(max_depth= val, random_state= 42), x, y, cv= kf, scoring=\"neg_mean_squared_error\")\n",
        "    print(f'For max depth: {val}')\n",
        "    rmse(score.mean())\n",
        "For max depth: 1\n",
        "rmse= nan\n",
        "For max depth: 2\n",
        "rmse= nan\n",
        "For max depth: 3\n",
        "rmse= nan\n",
        "For max depth: 4\n",
        "rmse= nan\n",
        "For max depth: 5\n",
        "rmse= nan\n",
        "For max depth: 6\n",
        "rmse= nan\n",
        "For max depth: 7\n",
        "rmse= nan\n",
        "For max depth: 8\n",
        "rmse= nan\n",
        "For max depth: 9\n",
        "rmse= nan\n",
        "For max depth: 10\n",
        "rmse= nan\n",
        "C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:372: FitFailedWarning: \n",
        "10 fits failed out of a total of 10.\n",
        "The score on these train-test partitions for these parameters will be set to nan.\n",
        "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
        "\n",
        "Below are more details about the failures:\n",
        "--------------------------------------------------------------------------------\n",
        "9 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Presence'\n",
        "\n",
        "--------------------------------------------------------------------------------\n",
        "1 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Absence'\n",
        "\n",
        "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
        "C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:372: FitFailedWarning: \n",
        "10 fits failed out of a total of 10.\n",
        "The score on these train-test partitions for these parameters will be set to nan.\n",
        "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
        "\n",
        "Below are more details about the failures:\n",
        "--------------------------------------------------------------------------------\n",
        "9 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Presence'\n",
        "\n",
        "--------------------------------------------------------------------------------\n",
        "1 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Absence'\n",
        "\n",
        "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
        "C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:372: FitFailedWarning: \n",
        "10 fits failed out of a total of 10.\n",
        "The score on these train-test partitions for these parameters will be set to nan.\n",
        "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
        "\n",
        "Below are more details about the failures:\n",
        "--------------------------------------------------------------------------------\n",
        "9 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Presence'\n",
        "\n",
        "--------------------------------------------------------------------------------\n",
        "1 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Absence'\n",
        "\n",
        "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
        "C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:372: FitFailedWarning: \n",
        "10 fits failed out of a total of 10.\n",
        "The score on these train-test partitions for these parameters will be set to nan.\n",
        "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
        "\n",
        "Below are more details about the failures:\n",
        "--------------------------------------------------------------------------------\n",
        "9 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Presence'\n",
        "\n",
        "--------------------------------------------------------------------------------\n",
        "1 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Absence'\n",
        "\n",
        "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
        "C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:372: FitFailedWarning: \n",
        "10 fits failed out of a total of 10.\n",
        "The score on these train-test partitions for these parameters will be set to nan.\n",
        "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
        "\n",
        "Below are more details about the failures:\n",
        "--------------------------------------------------------------------------------\n",
        "9 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Presence'\n",
        "\n",
        "--------------------------------------------------------------------------------\n",
        "1 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Absence'\n",
        "\n",
        "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
        "C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:372: FitFailedWarning: \n",
        "10 fits failed out of a total of 10.\n",
        "The score on these train-test partitions for these parameters will be set to nan.\n",
        "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
        "\n",
        "Below are more details about the failures:\n",
        "--------------------------------------------------------------------------------\n",
        "9 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Presence'\n",
        "\n",
        "--------------------------------------------------------------------------------\n",
        "1 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Absence'\n",
        "\n",
        "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
        "C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:372: FitFailedWarning: \n",
        "10 fits failed out of a total of 10.\n",
        "The score on these train-test partitions for these parameters will be set to nan.\n",
        "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
        "\n",
        "Below are more details about the failures:\n",
        "--------------------------------------------------------------------------------\n",
        "9 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Presence'\n",
        "\n",
        "--------------------------------------------------------------------------------\n",
        "1 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Absence'\n",
        "\n",
        "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
        "C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:372: FitFailedWarning: \n",
        "10 fits failed out of a total of 10.\n",
        "The score on these train-test partitions for these parameters will be set to nan.\n",
        "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
        "\n",
        "Below are more details about the failures:\n",
        "--------------------------------------------------------------------------------\n",
        "9 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Presence'\n",
        "\n",
        "--------------------------------------------------------------------------------\n",
        "1 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Absence'\n",
        "\n",
        "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
        "C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:372: FitFailedWarning: \n",
        "10 fits failed out of a total of 10.\n",
        "The score on these train-test partitions for these parameters will be set to nan.\n",
        "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
        "\n",
        "Below are more details about the failures:\n",
        "--------------------------------------------------------------------------------\n",
        "9 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Presence'\n",
        "\n",
        "--------------------------------------------------------------------------------\n",
        "1 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Absence'\n",
        "\n",
        "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
        "C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py:372: FitFailedWarning: \n",
        "10 fits failed out of a total of 10.\n",
        "The score on these train-test partitions for these parameters will be set to nan.\n",
        "If these failures are not expected, you can try to debug them by setting error_score='raise'.\n",
        "\n",
        "Below are more details about the failures:\n",
        "--------------------------------------------------------------------------------\n",
        "9 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Presence'\n",
        "\n",
        "--------------------------------------------------------------------------------\n",
        "1 fits failed with the following error:\n",
        "Traceback (most recent call last):\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\model_selection\\_validation.py\", line 680, in _fit_and_score\n",
        "    estimator.fit(X_train, y_train, **fit_params)\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 1320, in fit\n",
        "    X_idx_sorted=X_idx_sorted,\n",
        "  File \"C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\tree\\_classes.py\", line 227, in fit\n",
        "    y = np.ascontiguousarray(y, dtype=DOUBLE)\n",
        "ValueError: could not convert string to float: 'Absence'\n",
        "\n",
        "  warnings.warn(some_fits_failed_message, FitFailedWarning)\n",
        "score = cross_val_score(tree.DecisionTreeClassifier(random_state= 42), x, y, cv= kf, scoring=\"accuracy\")\n",
        "print(f'Scores for each fold are: {score}')\n",
        "print(f'Average score: {\"{:.2f}\".format(score.mean())}')\n",
        "Scores for each fold are: [0.81481481 0.7037037  0.77777778 0.77777778 0.74074074 0.7037037\n",
        " 0.85185185 0.81481481 0.74074074 0.62962963]\n",
        "Average score: 0.76\n",
        "def run_cross_validation_on_trees(X, y, tree_depths, cv=10, scoring='accuracy'):\n",
        "    cv_scores_list = []\n",
        "    cv_scores_std = []\n",
        "    cv_scores_mean = []\n",
        "    accuracy_scores = []\n",
        "    for depth in tree_depths:\n",
        "        tree_model = DecisionTreeClassifier(max_depth=depth)\n",
        "        cv_scores = cross_val_score(tree_model, X, y, cv=cv, scoring=scoring)\n",
        "        cv_scores_list.append(cv_scores)\n",
        "        cv_scores_mean.append(cv_scores.mean())\n",
        "        cv_scores_std.append(cv_scores.std())\n",
        "        accuracy_scores.append(tree_model.fit(X, y).score(X, y))\n",
        "    cv_scores_mean = np.array(cv_scores_mean)\n",
        "    cv_scores_std = np.array(cv_scores_std)\n",
        "    accuracy_scores = np.array(accuracy_scores)\n",
        "    return cv_scores_mean, cv_scores_std, accuracy_scores\n",
        "\n",
        "def plot_cross_validation_on_trees(depths, cv_scores_mean, cv_scores_std, accuracy_scores, title):\n",
        "    fig, ax = plt.subplots(1,1, figsize=(25,5))\n",
        "    ax.plot(depths, cv_scores_mean, '-o', label='mean cross-validation accuracy', alpha=0.9)\n",
        "    ax.fill_between(depths, cv_scores_mean-2*cv_scores_std, cv_scores_mean+2*cv_scores_std, alpha=0.2)\n",
        "    ylim = plt.ylim()\n",
        "    ax.plot(depths, accuracy_scores, '-*', label='train accuracy', alpha=0.9)\n",
        "    ax.set_title(title, fontsize=16)\n",
        "    ax.set_xlabel('Tree depth', fontsize=14)\n",
        "    ax.set_ylabel('Accuracy', fontsize=14)\n",
        "    ax.set_ylim(ylim)\n",
        "    ax.set_xticks(depths)\n",
        "    ax.legend()\n",
        "\n",
        "sm_tree_depths = range(1,20)\n",
        "sm_cv_scores_mean, sm_cv_scores_std, sm_accuracy_scores = run_cross_validation_on_trees(X_train, y_train, sm_tree_depths)\n",
        "\n",
        "plot_cross_validation_on_trees(sm_tree_depths, sm_cv_scores_mean, sm_cv_scores_std, sm_accuracy_scores, \n",
        "                               'Accuracy per decision tree depth on training data')\n",
        "\n",
        "tree_model = DecisionTreeClassifier(max_depth=5,criterion='entropy')\n",
        "cv_scores = cross_val_score(tree_model, x, y, cv=10, scoring='accuracy')\n",
        "m=tree_model.fit(x, y)\n",
        "prediction=m.predict(X_test)\n",
        "cm= confusion_matrix(y_test,prediction)\n",
        "sns.heatmap(cm, annot=True,cmap='winter',linewidths=0.3, linecolor='black',annot_kws={\"size\": 20})\n",
        "print(classification_report(y_test, prediction))\n",
        "              precision    recall  f1-score   support\n",
        "\n",
        "     Absence       0.89      1.00      0.94        40\n",
        "    Presence       1.00      0.82      0.90        28\n",
        "\n",
        "    accuracy                           0.93        68\n",
        "   macro avg       0.94      0.91      0.92        68\n",
        "weighted avg       0.93      0.93      0.93        68\n",
        "\n",
        "\n",
        "TP=cm[0][0]\n",
        "TN=cm[1][1]\n",
        "FN=cm[1][0]\n",
        "FP=cm[0][1]\n",
        "print('Testing Accuracy for Decision Tree:',(TP+TN)/(TP+TN+FN+FP))\n",
        "print('Testing Sensitivity for Decision Tree:',(TP/(TP+FN)))\n",
        "print('Testing Specificity for Decision Tree:',(TN/(TN+FP)))\n",
        "print('Testing Precision for Decision Tree:',(TP/(TP+FP)))\n",
        "Testing Accuracy for Decision Tree: 0.9264705882352942\n",
        "Testing Sensitivity for Decision Tree: 0.8888888888888888\n",
        "Testing Specificity for Decision Tree: 1.0\n",
        "Testing Precision for Decision Tree: 1.0\n",
        "idx_max = sm_cv_scores_mean.argmax()\n",
        "sm_best_tree_depth = sm_tree_depths[idx_max]\n",
        "sm_best_tree_cv_score = sm_cv_scores_mean[idx_max]\n",
        "sm_best_tree_cv_score_std = sm_cv_scores_std[idx_max]\n",
        "print('The depth-{} tree achieves the best mean cross-validation accuracy {} on training dataset'.format(\n",
        "      sm_best_tree_depth, round(sm_best_tree_cv_score*100,5)))\n",
        "The depth-3 tree achieves the best mean cross-validation accuracy 82.14286 on training dataset\n",
        "predictiontree=tree_model.predict(X_test)\n",
        "from sklearn.tree import DecisionTreeClassifier\n",
        "\n",
        "dtclas=DecisionTreeClassifier()\n",
        "modeldt=dtclas.fit(X_train,y_train)\n",
        "predictiondt=modeldt.predict(X_test)\n",
        "cmdt= confusion_matrix(y_test,predictiondt)\n",
        "sns.heatmap(cmdt, annot=True,cmap='winter')\n",
        "print(classification_report(y_test, predictiondt))\n",
        "              precision    recall  f1-score   support\n",
        "\n",
        "     Absence       0.77      0.75      0.76        40\n",
        "    Presence       0.66      0.68      0.67        28\n",
        "\n",
        "    accuracy                           0.72        68\n",
        "   macro avg       0.71      0.71      0.71        68\n",
        "weighted avg       0.72      0.72      0.72        68\n",
        "\n",
        "\n",
        "from sklearn.ensemble import RandomForestClassifier\n",
        "rfc=RandomForestClassifier(random_state=42)\n",
        "param_grid = { \n",
        "    'n_estimators': [200, 500],\n",
        "    'max_features': ['auto', 'sqrt', 'log2'],\n",
        "    'max_depth' : [4,5,6,7,8],\n",
        "    'criterion' :['gini', 'entropy']\n",
        "}\n",
        "CV_rfc = GridSearchCV(estimator=rfc, param_grid=param_grid, cv= 5)\n",
        "CV_rfc.fit(X_train, y_train)\n",
        "GridSearchCV(cv=5, estimator=RandomForestClassifier(random_state=42),\n",
        "             param_grid={'criterion': ['gini', 'entropy'],\n",
        "                         'max_depth': [4, 5, 6, 7, 8],\n",
        "                         'max_features': ['auto', 'sqrt', 'log2'],\n",
        "                         'n_estimators': [200, 500]})\n",
        "CV_rfc.best_params_\n",
        "{'criterion': 'gini',\n",
        " 'max_depth': 5,\n",
        " 'max_features': 'auto',\n",
        " 'n_estimators': 500}\n",
        "from sklearn.linear_model import LogisticRegression\n",
        "lr=LogisticRegression(C=1.0, class_weight='balanced', dual=False,\n",
        "                   fit_intercept=True, intercept_scaling=1, l1_ratio=None,\n",
        "                   max_iter=100, multi_class='auto', n_jobs=None, penalty='l2',\n",
        "                   random_state=1234, solver='lbfgs', tol=0.0001, verbose=0,\n",
        "                   warm_start=False)\n",
        "model1=lr.fit(X_train,y_train)\n",
        "C:\\Users\\Basker\\AppData\\Local\\Programs\\Python\\Python37\\lib\\site-packages\\sklearn\\linear_model\\_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):\n",
        "STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.\n",
        "\n",
        "Increase the number of iterations (max_iter) or scale the data as shown in:\n",
        "    https://scikit-learn.org/stable/modules/preprocessing.html\n",
        "Please also refer to the documentation for alternative solver options:\n",
        "    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression\n",
        "  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,\n",
        "from sklearn.metrics import accuracy_score\n",
        "input=(63,1,3,145,200,150,98,0,0,0,0,0,0)\n",
        "input_as_numpy=np.asarray(input)\n",
        "input_reshaped=input_as_numpy.reshape(1,-1)\n",
        "pre1=tree_model.predict(input_reshaped)\n",
        "print(pre1)\n",
        "a1 = accuracy_score(pre1,model1.predict(input_reshaped)) * 100\n",
        "print(a1)\n",
        "['Absence']\n",
        "100.0\n",
        "from sklearn.metrics import accuracy_score\n",
        "input=(70,1,4,130,322,0,2,109,0,2.4,2,3,3)\n",
        "input_as_numpy=np.asarray(input)\n",
        "input_reshaped=input_as_numpy.reshape(1,-1)\n",
        "pre1=tree_model.predict(input_reshaped)\n",
        "print(pre1)\n",
        "a1 = accuracy_score(pre1,model1.predict(input_reshaped)) * 100\n",
        "print(a1)\n",
        "['Presence']\n",
        "100.0\n",
        " \n",
        " "
      ]
    }
  ]
}